### PR TITLE
Replicated reads and writes using Consistent Hashing

### LINKED ISSUE #21: Replication support - Part 1
- [x] Add support for getting multiple nodes from CHRing. - #23 
- [x] Replication Interceptor for doing synchronous replication as writes happen. - #23 and #27

### PR DESCRIPTION
**Pull request by ashwanthkumar**
- CHRing#find(key, n) - returns n nodes from the ring closest to the key - can contain duplicates
- CHRing#findNUnique(key, n) - returns n unique nodes from the right closest to the key
- Changed contract of the Partitioner to return MemberAddress instead of VNode, since VNode is CHRing specific detail
- Added few more tests around CHPartitioner and CHRing

Part of #21

### CODE REVIEW

File: src/main/scala/in/ashwanthkumar/suuchi/partitioner/ConsistentHashRing.scala [hunk 1]
```diff
@@ -28,16 +36,75 @@ class ConsistentHashRing(hashFn: Hash, vnodeFactor: Int = 3) {
...
+   * This returns the closest n nodes in order for the object. There may be
+   * duplicates.
+   */
+  def find(key: Array[Byte], n: Int) = {
+    if (sortedMap.isEmpty) Nil
+    else {
+      var hash = hashFn.hash(key)
+      (0 until n).map { index =>
+        val (newHash, candidate) = findCandidate(hash)
+        hash = newHash + 1
```
Comment (brewkode) [inline src/main/scala/in/ashwanthkumar/suuchi/partitioner/ConsistentHashRing.scala]: Saw the reference and this as well :)

**Comment by brewkode** [comment]
Lets catch up on this before I merge. Hope thats ok.

**Comment by ashwanthkumar** [comment]
@brewkode Shall I go ahead and merge this?

File: src/test/scala/in/ashwanthkumar/suuchi/router/RouterSpec.scala [hunk 2]
```diff
@@ -11,29 +11,29 @@ class NeverRoute extends RoutingStrategy {
   /**
    * @inheritdoc
    */
-  override def route[ReqT]: PartialFunction[ReqT, Option[MemberAddress]] = PartialFunction.empty
+  override def route[ReqT]: PartialFunction[ReqT, List[MemberAddress]] = PartialFunction.empty
 }
 
-class RouterTest extends FlatSpec {
+class RouterSpec extends FlatSpec {
```
Comment (ashwanthkumar) [inline src/test/scala/in/ashwanthkumar/suuchi/router/RouterSpec.scala]: Should we rename this to `HandleOrForwardRouterSpec` ?

File: src/main/scala/in/ashwanthkumar/suuchi/router/HandleOrForwardRouter.scala [hunk 3]
```diff
@@ -27,17 +27,21 @@ class Router(routingStrategy: RoutingStrategy, self: MemberAddress) extends Serv
         // TODO - Handle forwarding loop here
         if(routingStrategy.route.isDefinedAt(incomingRequest)) {
           routingStrategy route incomingRequest match {
-            case Some(node) if !node.equals(self) =>
-              log.trace(s"Forwarding request to $node")
+            case items if items.nonEmpty && !items.exists(_.equals(self)) =>
+              val node = items.head
```
Comment (ashwanthkumar) [inline src/main/scala/in/ashwanthkumar/suuchi/router/HandleOrForwardRouter.scala]: Somehow forwarding to just the `items.head` - feels a little weird.

**Comment by ashwanthkumar** [comment]
Most of my comments are addressed as part of #27

**Comment by brewkode** [comment]
@ashwanthkumar Does this need to be merged?

**Comment by ashwanthkumar** [comment]
@brewkode #27 was merge into this branch so yeah we need to merge this too.

**Comment by ashwanthkumar** [comment]
@brewkode incorporated the feedback comments from #27 in here. Have one final look at the changes and we can merge it in.

**Comment by brewkode** [comment]
@ashwanthkumar Merging this!

**Pull request merged**